gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -96,26 +93,15 @@ public GoogleCloudStorageGrpcWriteChannel(
 
   @Override
   public void setDirectUploadEnabled(boolean enableDirectUpload) {
-    // TODO(b/135137933): Figure out if we need to support non-resumable uploads.
-    // It doesn't look as if the connector uses it at all, but there's code
-    // to support it, so maybe I'm missing something.
     if (enableDirectUpload) {
       throw new UnsupportedOperationException("Direct upload not supported.");
     }
   }
 
   @Override
   public void setUploadChunkSize(int chunkSize) {
-    Preconditions.checkArgument(chunkSize > 0, "Upload chunk size must be great than 0.");
-    Preconditions.checkArgument(
-        chunkSize % GCS_MINIMUM_CHUNK_SIZE == 0,
-        "Upload chunk size must be a multiple of GCS_MINIMUM_CHUNK_SIZE");
-    if ((chunkSize > GCS_UPLOAD_GRANULARITY) && (chunkSize % GCS_UPLOAD_GRANULARITY != 0)) {
-      logger.atWarning().log(
-          "Upload chunk size should be a multiple of %s for best performance, got %s",
-          GCS_UPLOAD_GRANULARITY, chunkSize);
-    }
-    uploadChunkSize = chunkSize;
+    // No-op for GCS gRPC - we always upload using a stream of MAX_WRITE_CHUNK_BYTES sized messages.
```

Comment:
Is there a possibility that optimal chunk size will change in the future, and it will be handy to have an option to adjust it in the client?

Reply:
The optimal performance will come from sending messages whose size is the max allowed by GCS (2 MiB).

I think chunk size probably matters more for other HDFS cases, where you're dealing with different write quantum alignment requirements for different underlying FS implementations. With GCS, what you want is to write in some integral multiple of 512 KiB, and you want to do it in the fewest number of largest allowed messages - so, 2 MiB makes sense.

If something about that reasoning changes we could always change the implementation.

Aside: In general I prefer to add flexibility for such configuration once there's a concrete reason to do so, not because there's a possibility that it might be useful in the future.

Reply:
I think that we added this option in the first place because at some point optimal write chunk was 8 MiB for GCS but later it become 2 MiB, so seems like there are a real possibility that it could change in the future.

Reply:
I agree it can change over time, but I would suggest the way we deal with that is to update this implementation.
I don't see value in letting customers control it. They may not understand the implications and could set an inappropriate value and then complain when they get poor performance, and we'll get an escalation to investigate.

Reply:
I think direct upload implemented with `uploadType=multipart` parameter in JSON API:
https://cloud.google.com/storage/docs/json_api/v1/how-tos/upload

Essentially upload when you don't need to create a separate resumable upload session - it saves at least one round-trip.

If this is/will be supported in gRPC API too, then I think we should have this TODO and address it before GA.

Reply:
The problem is that as soon as we ship connector in Dataproc image it's impossible to change binaries but we can change configuration if there will be a need, so we usually opt-in to have an option to adjust configuration.

Customer already can change a lot of options that will negatively impact performance, so I think this is a secondary concern.

In this case, if you are confident that we will not need to change this in foreseeable future (at least couple years) than I'm OK to leave it as is.

Reply:
I've only seen the optimal upload chunk size change once in the 10 years I've been on the GCS project ;-)